detection_rules/version_lock.py

```diff
@@ -0,0 +1,164 @@
+# Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+# or more contributor license agreements. Licensed under the Elastic License
+# 2.0; you may not use this file except in compliance with the Elastic License
+# 2.0.
+"""Helper utilities to manage the version lock."""
+from copy import deepcopy
+from typing import List, Optional
+
+import click
+
+from .rule import TOMLRule
+from .semver import Version
+from .utils import dict_hash, load_etc_dump, save_etc_dump, cached
+
+ETC_VERSION_LOCK_FILE = "version.lock.json"
+ETC_DEPRECATED_RULES_FILE = "deprecated_rules.json"
+MIN_LOCK_VERSION_DEFAULT = Version("7.13.0")
+
+
+def _convert_lock_version(stack_version: Optional[str]) -> Version:
+    """Convert an optional stack version to the minimum for the lock."""
+    if stack_version is None:
+        return MIN_LOCK_VERSION_DEFAULT
+    return max(Version(stack_version), MIN_LOCK_VERSION_DEFAULT)
+
+
+@cached
+def get_locked_version(rule_id: str, min_stack_version: Optional[str] = None) -> Optional[int]:
+    rules_versions = load_versions()
+
+    if rule_id in rules_versions:
+        latest_version_info = rules_versions[rule_id]
+        stack_version_info = latest_version_info.get("previous", {}).get(min_stack_version, latest_version_info)
```

Comment:
can you paste a sample of what the new structure will look like for the version.lock file

Reply:
```javascript
"id": {
  "min_stack_version": "7.15.0",
  "rule_name: "Some Rule"
  "version": 10
  "sha256": "...",
  "previous": {
    "7.13.0": {
      "rule_name: "Some Rule"
      "sha256": "...",
      "version": 5
    },
    "7.14.0": {
      "min_stack_version": "7.15.0",
      "rule_name: "Some Rule"
      "sha256": "...",
      "version": 7
    }
  }
}